pkg/detectors/alibaba/alibaba.go

```diff
@@ -87,49 +107,75 @@ func (s Scanner) FromData(ctx context.Context, verify bool, data []byte) (result
-				params.Add("SignatureVersion", "1.0")
-				params.Add("Timestamp", dateISO)
-				params.Add("Version", "2014-05-26")
-
-				stringToSign := buildStringToSign(req.Method, params.Encode())
-				signature := GetSignature(stringToSign, resMatch+"&") // Get Signature HMAC SHA1
-				params.Add("Signature", signature)
-				req.URL.RawQuery = params.Encode()
-
-				req.Header.Add("Content-Type", "text/xml;charset=utf-8")
-				req.Header.Add("Content-Length", strconv.Itoa(len(params.Encode())))
-				res, err := client.Do(req)
-				if err == nil {
-					defer res.Body.Close()
-					if res.StatusCode >= 200 && res.StatusCode < 300 {
-						s1.Verified = true
-					} else {
-						// This function will check false positives for common test words, but also it will make sure the key appears 'random' enough to be a real key.
-						if detectors.IsKnownFalsePositive(resMatch, detectors.DefaultFalsePositives, true) {
-							continue
-						}
-					}
-				}
+				client := s.getClient()
+				isVerified, verificationErr := verifyAlibaba(ctx, client, resIdMatch, resMatch)
+				s1.Verified = isVerified
+				s1.SetVerificationError(verificationErr, resMatch)
 			}
 
+			// This function will check false positives for common test words, but also it will make sure the key appears 'random' enough to be a real key.
+			if !s1.Verified && detectors.IsKnownFalsePositive(resMatch, detectors.DefaultFalsePositives, true) {
+				continue
+			}
 			results = append(results, s1)
 		}
 	}
 
 	return results, nil
 }
 
+func verifyAlibaba(ctx context.Context, client *http.Client, resIdMatch, resMatch string) (bool, error) {
+	req, err := http.NewRequestWithContext(ctx, "GET", "https://ecs.aliyuncs.com/?", nil)
+	if err != nil {
+		return false, err
+	}
+
+	dateISO := time.Now().UTC().Format("2006-01-02T15:04:05Z07:00")
+	params := req.URL.Query()
+	params.Add("AccessKeyId", resIdMatch)
+	params.Add("Action", "DescribeRegions")
+	params.Add("Format", "JSON")
+	params.Add("SignatureMethod", "HMAC-SHA1")
+	params.Add("SignatureNonce", randString(16))
+	params.Add("SignatureVersion", "1.0")
+	params.Add("Timestamp", dateISO)
+	params.Add("Version", "2014-05-26")
+
+	stringToSign := buildStringToSign(req.Method, params.Encode())
+	signature := GetSignature(stringToSign, resMatch+"&") // Get Signature HMAC SHA1
+	params.Add("Signature", signature)
+	req.URL.RawQuery = params.Encode()
+
+	req.Header.Add("Content-Type", "text/xml;charset=utf-8")
+	req.Header.Add("Content-Length", strconv.Itoa(len(params.Encode())))
+
+	res, err := client.Do(req)
+	if err != nil {
+		return false, err
+	}
+	defer res.Body.Close()
+
+	var alibabaResp alibabaResp
+	if err = json.NewDecoder(res.Body).Decode(&alibabaResp); err != nil {
+		return false, err
+	}
+
+	if res.StatusCode == http.StatusOK {
+		return true, nil
+	} else if res.StatusCode < http.StatusBadRequest || res.StatusCode > http.StatusNotFound {
+		// 400 used for most of error cases
+		// 404 used if the AccessKeyId is not valid
```

Comment:
what about 401, 402, and 403?

Reply:
i didn't see 401, 402, and 403 in the my testing. but it was easier to add the range

Reply:
i see two problems with doing the "easier" thing here. the smaller one is that you're assuming that the external system will continue to behave in the way you've seen it behave, which is always a risk. now in this particular case, that's unlikely to cause real problems - 401s and 403s are probably determine verification errors if they ever do show up, and 402s will probably never happen - but it's not a great happen to get into.

the bigger problem is that your code doesn't match your comment, so the next person to read this is going to be confused and probably ask the same question i did. but you won't be around to answer it! so they'll have to spend time chasing down a ghost.

Reply:
fair